subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -80,38 +90,22 @@ ProviderInternal<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter)
                 .build();
         }
 
-        return new DefaultProvider<>(() -> {
-            if (!filterInternal.isConfigured()) {
-                return null;
-            }
+        if (!filterInternal.isConfigured()) {
+            return null;
+        }
 
-            synchronized (matchingToolchains) {
-                if (matchingToolchains.containsKey(filterInternal.toKey())) {
-                    return handleMatchingToolchainCached(filterInternal);
-                } else {
-                    return handleMatchingToolchainUnknown(filterInternal);
-                }
+        Object resolutionResult = matchingToolchains.computeIfAbsent(filterInternal.toKey(), key -> {
```

Comment:
In order to drop the synchronization here, doesn't `matchingToolchains` need to be a ConcurrentHashMap` and not just a `HashMap`?

Reply:
Good catch! I was sure it was a concurrent map before, but it seems that has changed after the download SPI has been merged.